programs/stake/src/stake_state.rs

```diff
@@ -842,6 +823,99 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+struct ValidatedSplitInfo {
+    source_remaining_balance: u64,
+    destination_rent_exempt_reserve: u64,
+}
+
+/// Ensure the split amount is valid.  This checks:
+/// 1. Both the source and destination accounts meet the minimum balance requirements, which is the
+///    rent exempt reserve plus the minimum stake delegation.
+/// 2. The source account has enough lamports for the requested split amount.
+/// 3. The destination ends up with enough lamports to meet the minimum stake delegation.
+/// If not, return an error.
+fn validate_split_amount(
+    source_account: &KeyedAccount,
+    destination_account: &KeyedAccount,
+    lamports: u64,
+    source_meta: &Meta,
+) -> Result<ValidatedSplitInfo, InstructionError> {
+    let source_lamports = source_account.lamports()?;
+
+    // Split amount has to be something
+    if lamports == 0 {
+        return Err(InstructionError::InsufficientFunds);
+    }
+
+    // Obviously cannot split more than what the source account has
+    if lamports > source_lamports {
+        return Err(InstructionError::InsufficientFunds);
+    }
+
+    // Verify that the source account still has enough lamports left after splitting:
+    // EITHER at least the minimum balance, OR zero (in this case the source
+    // account is transferring all lamports to new destination account, and the source
+    // account will be closed)
+    let source_minimum_balance = source_meta
+        .rent_exempt_reserve
+        .saturating_add(MINIMUM_STAKE_DELEGATION);
+    let source_remaining_balance = source_lamports.saturating_sub(lamports);
+    if source_remaining_balance == 0 {
+        // full amount is a withdrawal
+        // nothing to do here
+    } else if source_remaining_balance < source_minimum_balance {
+        // the remaining balance is too low to do the split
+        return Err(InstructionError::InsufficientFunds);
+    } else {
+        // all clear!
+        // nothing to do here
+    }
+
+    // Verify the destination account also meets the minimum balance requirements
+    // This must account for:
+    // 1. The destination account having a different rent exempt reserve due to data size changes.
+    // 2. The destination account being prefunded, which would lower the minimum split amount.
+    // 3. The destination account must have enough lamports to meet the minimum stake delegation,
+    //    which would raise the minimum split amount.
+    let destination_rent_exempt_reserve = calculate_split_rent_exempt_reserve(
+        source_meta.rent_exempt_reserve,
+        source_account.data_len()? as u64,
+        destination_account.data_len()? as u64,
+    );
+    let destination_minimum_balance =
+        destination_rent_exempt_reserve.saturating_add(MINIMUM_STAKE_DELEGATION);
+    let destination_balance_deficit =
+        destination_minimum_balance.saturating_sub(destination_account.lamports()?);
+    let minimum_split_amount = std::cmp::max(MINIMUM_STAKE_DELEGATION, destination_balance_deficit);
```

Comment:
Sorry, I don't think I made the contra to this part clear.  There could be a different strange behavior with this, where anytime I split an `Initialized` stake into an account that already has `destination_minimum_balance`, I can't split anything less than `MINIMUM_STAKE_DELEGATION`, which is fine for now, but once the number is big, it means I'll have to do a `withdraw` + `transfer`, which is probably fine, but something to keep in mind.

The best of all behaviors would be to take the `max` only if `source_account` is a `Stake`, which can be expressed through an additional `is_stake` flag or something.  Does that make sense?

Reply:
Ah, OK. I glossed over that in your other comment, sorry! I've just pushed another commit that hopefully fixes this.

I put the changes inside the match within `split()`, since I wanted to try and keep `validate_split_amount()` to be state-agnostic. I also did _not_ add a check if splitting 100%, since I'm making the assumption that the source account meets the minimum requirements.

(But note, then what happens if the MINIMUM_STAKE_DELEGATION is raised and there's an existing stake account that tries to split 100%. Should that succeed? Should that fail? I'm not sure. Luckily that doesn't need to be answered immediately for this PR.)

Reply:
> (But note, then what happens if the MINIMUM_STAKE_DELEGATION is raised and there's an existing stake account that tries to split 100%. Should that succeed? Should that fail? I'm not sure. Luckily that doesn't need to be answered immediately for this PR.)

Hehe indeed not for this PR, but my thinking was that only merge / deactivate / withdraw would be supported for existing stakes underneath `MINIMUM_STAKE_DELEGATION`.  You could even talk me out of "merge"